{"cells":[{"cell_type":"markdown","id":"f6767c73-18e4-4351-8b86-11715f927a22","metadata":{"id":"f6767c73-18e4-4351-8b86-11715f927a22"},"source":["# SV Hackathon \n","### October 9-12, 2022\n","### Current update on Oct. 10 12:41pm CT\n","### NISC Genomics Features Group\n"]},{"cell_type":"markdown","id":"cc28ff80-8995-4e08-9c0b-4ab1f1b02f71","metadata":{"id":"cc28ff80-8995-4e08-9c0b-4ab1f1b02f71"},"source":["### Starting with S1: (Updating GRCh38 resources to T2T-CHM13v2.0 reference) Generate T2T-CHM13v2.0 stratification BED file similar to what exists for GRCh38\n","- Running GEM on CHRM13v.2.0\n","- Testing it out first  on chr21, then maybe moving on to whole genome\n"]},{"cell_type":"markdown","id":"cee7cbf9-6cbf-45ad-ab7f-90e42f7500fa","metadata":{"id":"cee7cbf9-6cbf-45ad-ab7f-90e42f7500fa"},"source":["### Setting up our environment and tools"]},{"cell_type":"code","execution_count":null,"id":"cd54e5fa-201a-43ef-912c-b442a168f00b","metadata":{"id":"cd54e5fa-201a-43ef-912c-b442a168f00b"},"outputs":[],"source":["export MAIN=/path/to/\n","# This was all done on NIH's Biowulf Cluster"]},{"cell_type":"markdown","id":"114f72e6-c352-4708-a604-c623dbeedb92","metadata":{"id":"114f72e6-c352-4708-a604-c623dbeedb92"},"source":["Upload CHM13v2.0 fasta from locally downloaded source"]},{"cell_type":"code","execution_count":null,"id":"3c24fe5a-49f9-4dc5-9e75-4aee8fc00708","metadata":{"id":"3c24fe5a-49f9-4dc5-9e75-4aee8fc00708"},"outputs":[],"source":["# chm13v2.0.fa.gz\n","# chm13v2.0.fa.gz.gzi\n"]},{"cell_type":"markdown","id":"9d81fe95-b57b-4be6-9619-48f8a3f507f9","metadata":{"id":"9d81fe95-b57b-4be6-9619-48f8a3f507f9"},"source":["Extract chr 21 for testing"]},{"cell_type":"code","execution_count":null,"id":"c1346565-9c39-4273-9491-2cbdae8a70a7","metadata":{"id":"c1346565-9c39-4273-9491-2cbdae8a70a7","outputId":"4602abeb-ebce-4ea1-dc2e-c14ea2838a72"},"outputs":[{"name":"stdout","output_type":"stream","text":["[+] Loading samtools 1.15.1  ... \n"]}],"source":["module load samtools\n","samtools faidx chm13v2.0.fa.gz chr21 > chm13_chr21.fa.gz"]},{"cell_type":"code","execution_count":null,"id":"3d877311-0df5-4149-81f5-b3c84932d545","metadata":{"id":"3d877311-0df5-4149-81f5-b3c84932d545"},"outputs":[],"source":["samtools faidx chm13_chr21.fa.gz"]},{"cell_type":"markdown","id":"60869afa-e100-46cf-b7c8-5051ece149a5","metadata":{"id":"60869afa-e100-46cf-b7c8-5051ece149a5"},"source":["Make Hackathon directory for all fasta and output files"]},{"cell_type":"code","execution_count":null,"id":"354875fe-65fc-4b09-822f-47b65c8ab1ee","metadata":{"id":"354875fe-65fc-4b09-822f-47b65c8ab1ee"},"outputs":[],"source":["mkdir $MAIN/sv_hackathon/\n","mv chm13* $MAIN/sv_hackathon/"]},{"cell_type":"markdown","id":"b7be38a6-6786-431d-8d92-df8667b085e6","metadata":{"id":"b7be38a6-6786-431d-8d92-df8667b085e6"},"source":["Upload and install bedops binaries"]},{"cell_type":"code","execution_count":null,"id":"ec52870d-702e-4aea-93d4-2614890f94e1","metadata":{"collapsed":true,"jupyter":{"outputs_hidden":true},"tags":[],"id":"ec52870d-702e-4aea-93d4-2614890f94e1","outputId":"2b9e1578-bfb9-4948-eddb-678d7dc22fa0"},"outputs":[{"name":"stdout","output_type":"stream","text":["bin/\n","bin/bam2bed-typical\n","bin/bam2bed-megarow\n","bin/bam2bed-float128\n","bin/bam2bed_gnuParallel-typical\n","bin/bam2bed_gnuParallel-megarow\n","bin/bam2bed_gnuParallel-float128\n","bin/bam2bed_sge-typical\n","bin/bam2bed_sge-megarow\n","bin/bam2bed_sge-float128\n","bin/bam2bed_slurm-typical\n","bin/bam2bed_slurm-megarow\n","bin/bam2bed_slurm-float128\n","bin/bam2starch-typical\n","bin/bam2starch-megarow\n","bin/bam2starch-float128\n","bin/bam2starch_gnuParallel-typical\n","bin/bam2starch_gnuParallel-megarow\n","bin/bam2starch_gnuParallel-float128\n","bin/bam2starch_sge-typical\n","bin/bam2starch_sge-megarow\n","bin/bam2starch_sge-float128\n","bin/bam2starch_slurm-typical\n","bin/bam2starch_slurm-megarow\n","bin/bam2starch_slurm-float128\n","bin/gff2bed-typical\n","bin/gff2bed-megarow\n","bin/gff2bed-float128\n","bin/gff2starch-typical\n","bin/gff2starch-megarow\n","bin/gff2starch-float128\n","bin/gtf2bed-typical\n","bin/gtf2bed-megarow\n","bin/gtf2bed-float128\n","bin/gtf2starch-typical\n","bin/gtf2starch-megarow\n","bin/gtf2starch-float128\n","bin/gvf2bed-typical\n","bin/gvf2bed-megarow\n","bin/gvf2bed-float128\n","bin/gvf2starch-typical\n","bin/gvf2starch-megarow\n","bin/gvf2starch-float128\n","bin/psl2bed-typical\n","bin/psl2bed-megarow\n","bin/psl2bed-float128\n","bin/psl2starch-typical\n","bin/psl2starch-megarow\n","bin/psl2starch-float128\n","bin/rmsk2bed-typical\n","bin/rmsk2bed-megarow\n","bin/rmsk2bed-float128\n","bin/rmsk2starch-typical\n","bin/rmsk2starch-megarow\n","bin/rmsk2starch-float128\n","bin/sam2bed-typical\n","bin/sam2bed-megarow\n","bin/sam2bed-float128\n","bin/sam2starch-typical\n","bin/sam2starch-megarow\n","bin/sam2starch-float128\n","bin/vcf2bed-typical\n","bin/vcf2bed-megarow\n","bin/vcf2bed-float128\n","bin/vcf2starch-typical\n","bin/vcf2starch-megarow\n","bin/vcf2starch-float128\n","bin/wig2bed-typical\n","bin/wig2bed-megarow\n","bin/wig2bed-float128\n","bin/wig2starch-typical\n","bin/wig2starch-megarow\n","bin/wig2starch-float128\n","bin/starchcluster_sge-typical\n","bin/starchcluster_gnuParallel-typical\n","bin/starchcluster_slurm-typical\n","bin/starch-diff-typical\n","bin/starchcluster_sge-megarow\n","bin/starchcluster_gnuParallel-megarow\n","bin/starchcluster_slurm-megarow\n","bin/starch-diff-megarow\n","bin/starchcluster_sge-float128\n","bin/starchcluster_gnuParallel-float128\n","bin/starchcluster_slurm-float128\n","bin/starch-diff-float128\n","bin/sort-bed-typical\n","bin/update-sort-bed-slurm-typical\n","bin/update-sort-bed-starch-slurm-typical\n","bin/update-sort-bed-migrate-candidates-typical\n","bin/bedops-typical\n","bin/closest-features-typical\n","bin/bedmap-typical\n","bin/bedextract-typical\n","bin/starch-typical\n","bin/unstarch-typical\n","bin/starchcat-typical\n","bin/starchstrip-typical\n","bin/convert2bed-typical\n","bin/sort-bed-megarow\n","bin/update-sort-bed-slurm-megarow\n","bin/update-sort-bed-starch-slurm-megarow\n","bin/update-sort-bed-migrate-candidates-megarow\n","bin/bedops-megarow\n","bin/closest-features-megarow\n","bin/bedmap-megarow\n","bin/bedextract-megarow\n","bin/starch-megarow\n","bin/unstarch-megarow\n","bin/starchcat-megarow\n","bin/starchstrip-megarow\n","bin/convert2bed-megarow\n","bin/sort-bed-float128\n","bin/update-sort-bed-slurm-float128\n","bin/update-sort-bed-starch-slurm-float128\n","bin/update-sort-bed-migrate-candidates-float128\n","bin/bedops-float128\n","bin/closest-features-float128\n","bin/bedmap-float128\n","bin/bedextract-float128\n","bin/starch-float128\n","bin/unstarch-float128\n","bin/starchcat-float128\n","bin/starchstrip-float128\n","bin/convert2bed-float128\n","bin/switch-BEDOPS-binary-type\n","bin/bam2bed\n","bin/bam2bed_gnuParallel\n","bin/bam2bed_sge\n","bin/bam2bed_slurm\n","bin/bam2starch\n","bin/bam2starch_gnuParallel\n","bin/bam2starch_sge\n","bin/bam2starch_slurm\n","bin/gff2bed\n","bin/gff2starch\n","bin/gtf2bed\n","bin/gtf2starch\n","bin/gvf2bed\n","bin/gvf2starch\n","bin/psl2bed\n","bin/psl2starch\n","bin/rmsk2bed\n","bin/rmsk2starch\n","bin/sam2bed\n","bin/sam2starch\n","bin/vcf2bed\n","bin/vcf2starch\n","bin/wig2bed\n","bin/wig2starch\n","bin/starchcluster_sge\n","bin/starchcluster_gnuParallel\n","bin/starchcluster_slurm\n","bin/starch-diff\n","bin/sort-bed\n","bin/update-sort-bed-slurm\n","bin/update-sort-bed-starch-slurm\n","bin/update-sort-bed-migrate-candidates\n","bin/bedops\n","bin/closest-features\n","bin/bedmap\n","bin/bedextract\n","bin/starch\n","bin/unstarch\n","bin/starchcat\n","bin/starchstrip\n","bin/convert2bed\n"]}],"source":["tar jxvf bedops_linux_x86_64-v2.4.41.tar.bz2"]},{"cell_type":"code","execution_count":null,"id":"95cb0120-5a31-4e16-8e7f-da2a0ca74ee0","metadata":{"collapsed":true,"jupyter":{"outputs_hidden":true},"tags":[],"id":"95cb0120-5a31-4e16-8e7f-da2a0ca74ee0","outputId":"b135b925-c710-406c-999e-60f165498a07"},"outputs":[{"name":"stdout","output_type":"stream","text":["mkdir: cannot create directory './bedops': File exists\n"]}],"source":["mkdir $MAIN/bedops\n","cp bin/* $MAIN/bedops/\n","rm -r $MAIN/bin/"]},{"cell_type":"markdown","id":"04cd41a6-5796-4b6a-9066-e647a94073c1","metadata":{"id":"04cd41a6-5796-4b6a-9066-e647a94073c1"},"source":["Upload and install gem binaries"]},{"cell_type":"code","execution_count":null,"id":"9d5ef2d6-43a2-46dd-bddb-d9109c39b46e","metadata":{"collapsed":true,"jupyter":{"outputs_hidden":true},"tags":[],"id":"9d5ef2d6-43a2-46dd-bddb-d9109c39b46e","outputId":"75acde31-95ce-442f-89e6-09a0b46405db"},"outputs":[{"name":"stdout","output_type":"stream","text":["GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-indexer_bwt-dna\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/transcriptome-2-genome\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-mappability\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-indexer\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-indexer_generate\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-mappability-retriever\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/External/\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/External/gemtools\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/External/LICENSE\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gemtools\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-rna-mapper\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-retriever\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-2-sam\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-indexer_fasta2meta+cont\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gtf-2-junctions\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-2-wig\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-2-gem\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-info\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/compute-transcriptome\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/LICENSE\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/splits-2-junctions\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/bin/gem-mapper\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/man/\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/man/gem-indexer.man\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/man/gem-mapper.man\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/man/gem-2-sam.man\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/man/gem-2-gem.man\n","GEM-binaries-Linux-x86_64-core_i3-20130406-045632/man/gem-mappability.man\n"]}],"source":["tar jxvf GEM-binaries-Linux-x86_64-core_i3-20130406-045632.tbz2"]},{"cell_type":"markdown","id":"b4cd4d6a-17c6-4218-a07e-1e4a5c939636","metadata":{"id":"b4cd4d6a-17c6-4218-a07e-1e4a5c939636"},"source":["All binaries and files are now in directory"]},{"cell_type":"markdown","id":"1ae00bad-18a7-43e9-92dc-386caecb24fc","metadata":{"id":"1ae00bad-18a7-43e9-92dc-386caecb24fc"},"source":["### Time to import GEM .sh script \n","https://github.com/genome-in-a-bottle/genome-stratifications/blob/master/GRCh38/mappability/run_GEM_mappability_GRCh38.sh"]},{"cell_type":"code","execution_count":null,"id":"1d96fba6-7876-4d55-b8e2-236f9f4841f6","metadata":{"id":"1d96fba6-7876-4d55-b8e2-236f9f4841f6"},"outputs":[],"source":["# vim gem_mappability.sh"]},{"cell_type":"markdown","id":"18b7aa8d-84b7-4256-978c-5860ee2fa1e2","metadata":{"id":"18b7aa8d-84b7-4256-978c-5860ee2fa1e2"},"source":["Run for chr21 \\\n","Run twice each time  \\\n","-l 100 -m 2 -e 1 \\\n","-l 250 -m 0 -e 0 "]},{"cell_type":"code","execution_count":null,"id":"5eee2c69-f188-4bb9-8ee3-8127a04b9b75","metadata":{"collapsed":true,"jupyter":{"outputs_hidden":true},"tags":[],"id":"5eee2c69-f188-4bb9-8ee3-8127a04b9b75","outputId":"019c316d-d1c9-4141-bb76-8b0307dc154f"},"outputs":[{"name":"stdout","output_type":"stream","text":["Running job ,  on biowulf.nih.gov\n","Reference index does not exist, indexing.\n","Welcome to GEM-indexer build 1.423 (beta) - (2013/04/01 01:02:13 GMT)\n"," (c) 2008-2013 Paolo Ribeca <paolo.ribeca@gmail.com>\n"," (c) 2010-2013 Santiago Marco Sola <santiagomsola@gmail.com>\n"," (c) 2010-2013 Leonor Frias Moya <leonor.frias@gmail.com>\n","For the terms of use, run the program with the option --show-license.\n","************************************************************************\n","* WARNING: this is a beta version, provided for testing purposes only; *\n","*          check for updates at <http://gemlibrary.sourceforge.net>.   *\n","************************************************************************\n","Creating sequence and location files... done.\n","Computing DNA BWT (likely to take long)... done.\n","Generating index (likely to take long)... done.\n","Cleaning up... done.\n","Mappability file does not exist, generating\n","Welcome to GEM-mappability build 1.315 (beta) - (2013/03/29 02:59:40 GMT)\n"," (c) 2008-2013 Paolo Ribeca <paolo.ribeca@gmail.com>\n"," (c) 2010-2013 Santiago Marco Sola <santiagomsola@gmail.com>\n"," (c) 2010-2013 Leonor Frias Moya <leonor.frias@gmail.com>\n","For the terms of use, run the program with the option --show-license.\n","************************************************************************\n","* WARNING: this is a beta version, provided for testing purposes only; *\n","*          check for updates at <http://gemlibrary.sourceforge.net>.   *\n","************************************************************************\n","Fatal error (gem-mappability.c:300,main)\n"," Unrecognized commandline option '/data/alvarezjerezp2/sv_hackathon//chm13_chr21_gemmap_l_m_e'\n","Usage:\n"," gem-mappability\n","  -I <index_prefix>                       (mandatory)\n","  -C|--emulate-complement                 (for indices lacking it)\n","  --granularity <number>                  (default=10000)\n","  -o <output_prefix>                      (mandatory)\n","  --output-line-width <number>            (default=70)\n","  -l <read_length>                        (mandatory)\n","  -t|--approximation-threshold <number>|'disable'\n","                                          (default: first multiple bin)\n","  --mismatch-alphabet <symbols>           (default=\"ACGT\")\n","  -m <max_mismatches>|<%_mismatches>      (default=-1.00)\n","  -e <max_edit_distance>|<%_differences>  (default=0.04)\n","  --max-big-indel-length <number>         (default=15)\n","  --min-matched-bases <number>|<%>        (default=0.80)\n","  -s|--strata-after-best <number>         (default=1)\n","  -T <thread_number>                      (default=8)\n","  --show-license                          (print license)\n","  -h|--help                               (print usage)\n","Welcome to GEM-2-wig build 1.423 (beta) - (2013/04/01 01:02:13 GMT)\n"," (c) 2008-2013 Paolo Ribeca <paolo.ribeca@gmail.com>\n"," (c) 2010-2013 Santiago Marco Sola <santiagomsola@gmail.com>\n"," (c) 2010-2013 Leonor Frias Moya <leonor.frias@gmail.com>\n","For the terms of use, run the program with the option --show-license.\n","************************************************************************\n","* WARNING: this is a beta version, provided for testing purposes only; *\n","*          check for updates at <http://gemlibrary.sourceforge.net>.   *\n","************************************************************************\n","Usage:\n"," gem-2-wig\n","  -I|--index <archive_file>        (mandatory)\n","  -i|--input <mappability_file>    (default=stdin)\n","  -o|--output <prefix>             (mandatory)\n","  --show-license                   (show license and exit)\n","  -h|--help                        (print usage and exit)\n","Fatal error: exception Failure(\"Invalid argument to option '-i'\")\n","sed: can't read /data/alvarezjerezp2/sv_hackathon//chm13_chr21_gemmap_l_m_e.wig: No such file or directory\n","sed: can't read /data/alvarezjerezp2/sv_hackathon//chm13_chr21_gemmap_l_m_e.sizes: No such file or directory\n","20221009T201123  1919 biowulf.nih.gov\n"]}],"source":["sbatch gem_mappability.sh \n","# Altered paths and -l -m -e diirectly in script"]},{"cell_type":"code","execution_count":null,"id":"67d67e89-518e-40b6-8d44-13ce1548b3e3","metadata":{"id":"67d67e89-518e-40b6-8d44-13ce1548b3e3"},"outputs":[],"source":["Output to move forward chm13_chr21_gemmap_l100_m2_e1* and chm13_chr21_gemmap_l250_m0_e0*"]},{"cell_type":"markdown","id":"45d370d9-922c-4a8c-b264-92adbc23d4f5","metadata":{"id":"45d370d9-922c-4a8c-b264-92adbc23d4f5"},"source":["Run for whole genome"]},{"cell_type":"code","execution_count":null,"id":"f81529bc-50c7-47db-b01b-914ba165bbfc","metadata":{"tags":[],"id":"f81529bc-50c7-47db-b01b-914ba165bbfc","outputId":"1a70aae1-74d4-4fb8-bf42-5cd719d8e93b"},"outputs":[{"name":"stdout","output_type":"stream","text":["49313461\n"]}],"source":["sbatch gem_mappability.sh \n","# Altered files for whole genome"]},{"cell_type":"markdown","id":"d083e1c5-1f88-4f2a-a3e2-e2f2e2242434","metadata":{"id":"d083e1c5-1f88-4f2a-a3e2-e2f2e2242434"},"source":["Script was giving issue with .gz whole genome fasta file, decompressed and that seems to work"]},{"cell_type":"code","execution_count":null,"id":"26ff5b45-a931-43c3-8cf0-6c3ddcd13eca","metadata":{"id":"26ff5b45-a931-43c3-8cf0-6c3ddcd13eca"},"outputs":[],"source":["gunzip $MAIN/sv_hackathon/chm13.fa.gz"]},{"cell_type":"code","execution_count":null,"id":"c6d27997-d45d-41f2-96d6-62ef53cff09d","metadata":{"collapsed":true,"jupyter":{"outputs_hidden":true},"tags":[],"id":"c6d27997-d45d-41f2-96d6-62ef53cff09d","outputId":"4bdd1300-4951-412f-dc17-ae6855492478"},"outputs":[{"name":"stdout","output_type":"stream","text":[">chr1 CP068277.2 Homo sapiens isolate CHM13 chromosome 1\n","Caccctaaaccctaacccctaaccctaaccctaaccctaaccctaaccctaacccctaaaccctaaccctaaccctaacc\n","ctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccct\n","aaccctaaccctaacccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaacccta\n","accctaaccctaaccctaaccctaaccctaaccctaaccctaacccaaccctaaccctaaccctaaccctaaccctaacc\n","ctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccct\n","aaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaa\n","ccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaacc\n","ctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccct\n","aaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaaccctaa\n"]}],"source":["head $MAIN/sv_hackathon/chm13.fa"]},{"cell_type":"markdown","id":"6bfb8989-4fe4-4c02-9ad9-f59a3c782881","metadata":{"id":"6bfb8989-4fe4-4c02-9ad9-f59a3c782881"},"source":["Need to next run the second script for union files"]},{"cell_type":"code","execution_count":null,"id":"fa1d8629-af5d-43e1-8571-f0bd0a047e38","metadata":{"id":"fa1d8629-af5d-43e1-8571-f0bd0a047e38"},"outputs":[],"source":["# vim mappability_sort_gem.sh\n","# Work in progress"]},{"cell_type":"code","execution_count":null,"id":"b2f26f9a-dbc7-4ccf-82a6-79ebccf6bb0f","metadata":{"id":"b2f26f9a-dbc7-4ccf-82a6-79ebccf6bb0f"},"outputs":[],"source":[]}],"metadata":{"kernelspec":{"display_name":"Bash","language":"bash","name":"bash"},"language_info":{"codemirror_mode":"shell","file_extension":".sh","mimetype":"text/x-sh","name":"bash"},"colab":{"provenance":[]}},"nbformat":4,"nbformat_minor":5}